internal/process/manager.go

```diff
@@ -0,0 +1,448 @@
+		    __u16 flags;
+		    // detected_flags is used to determine if this operation has ever been
+		    // performed. if `flags & IS_LISTEN_15001` is false but `detected_flags &
+		    // IS_LISTEN_15001` is true, that means real true, we do not need recheck.
+		    // but if `detected_flags & IS_LISTEN_15001` is false, that probably means
+		    // we haven't tested it and need to retest it.
+		    __u16 detected_flags;
+	*/
+	ID            uint64
+	IsInMesh      uint32
+	CgroupIP      [4]uint32
+	Flags         uint16
+	DetectedFlags uint16
+}
+
+type ProcessManager interface {
+	Run(stop chan struct{}) error
+	OnPodStatusChanged(ip string, isInMesh bool, isAmbient bool, isZtunnel bool) error
+}
+
+type podInfo struct {
+	ip        string
+	isInMesh  bool
+	isAmbient bool
+	isZtunnel bool
+}
+
+type processManager struct {
+	lock            sync.RWMutex
+	cgroupMountPath string
+	pidCgroupMap    map[uint32]uint64
+	cgroupIPMap     map[uint64]string
+	ipCgroupsMap    map[string]map[uint64]bool
+	podIPModeMap    map[string]podInfo
+}
+
+func NewProcessManager(cgroupMountPath string) (ProcessManager, error) {
+	if cgroupMountPath == "" {
+		p, err := linux.GetCgroup2MountPath(config.HostProc)
+		if err != nil {
+			return nil, err
+		}
+		if p == "" {
+			// try cgroupv1
+			p, err = linux.GetCgroupSystemdMountPath(config.HostProc)
+			if err != nil {
+				return nil, err
+			}
+			if p == "" {
+				return nil, fmt.Errorf("can not fetch cgroup mount path")
+			}
+		}
+		cgroupMountPath = p
+	}
+	return &processManager{
+		cgroupMountPath: cgroupMountPath,
+		pidCgroupMap:    make(map[uint32]uint64),
+		cgroupIPMap:     make(map[uint64]string),
+		ipCgroupsMap:    make(map[string]map[uint64]bool),
+		podIPModeMap:    make(map[string]podInfo),
+	}, nil
+}
+
+// if process's ips more than 1 v4 or 1 v6, we dont deal with it
+// it may be hostnetwork pod
+func isSupportedPodProcess(ips []net.Addr) bool {
+	if len(ips) > 2 {
+		return false
+	}
+	ipv4 := 0
+	for _, ip := range ips {
+		if !strings.Contains(ip.String(), ":") {
+			ipv4 += 1
+		}
+	}
+	return ipv4 == 1
+}
+
+func isInKube(cgroup string) bool {
+	return strings.Contains(cgroup, "kubepods")
+}
+
+func getProcessCgroup(pid uint32) (string, error) {
+	cgr, err := os.ReadFile(fmt.Sprintf("%s/%d/cgroup", config.HostProc, pid))
+	if err != nil {
+		return "", err
+	}
+	lines := strings.Split(strings.TrimSpace(string(cgr)), "\n")
+	cgline := lines[len(lines)-1]
+	cg := strings.Split(cgline, ":")
+	if len(cg) < 3 {
+		return "", fmt.Errorf("error cgroup found: %s", cgline)
+	}
+	return strings.Join(cg[2:], ":"), nil
+}
+
+func getProcessIps(pid uint32) ([]net.Addr, error) {
+	netpath := fmt.Sprintf("%s/%d/ns/net", config.HostProc, pid)
+	netns, err := ns.GetNS(netpath)
+	if err != nil {
+		return nil, err
+	}
+	addrs := make([]net.Addr, 0)
+	err = netns.Do(func(nn ns.NetNS) error {
+		ifaces, _ := net.Interfaces()
+		for _, iface := range ifaces {
+			if strings.Contains(iface.Name, "istio") {
```

Comment:
Is this check for the geneve tunnel(`pistioin`/`pistioout`)?

Reply:
yes, this doesn't seem very elegant here, let me think about how to optimize this.